File: src/graph/runtimeGraphReconstruction.ts
import { RuntimeGraphContract, type RuntimeGraph } from "../contracts/graphContracts.js";
import { reconstructGraphFromSources } from "../reconstruction/reconstructGraph.js";
import { graphFingerprint } from "./runtimeGraph.js";

export function reconstructGraphFromIr(sources: Record<string, unknown>): {
  graph: RuntimeGraph;
  fingerprint: string;
  bounded: boolean;
} {
  const graph = reconstructGraphFromSources(sources);
  return { graph, fingerprint: graphFingerprint(graph), bounded: true };
}

export function rebuildGraphFromPartial(partial: RuntimeGraph): RuntimeGraph {
  const normalized = RuntimeGraphContract.normalize(partial);
  const nodes = normalized.nodes.map((n, i) => ({
    ...n,
    id: n.id ?? `rebuilt:${i}`,
  }));
  return RuntimeGraphContract.normalize({ nodes, edges: normalized.edges, bounded: true });
}

File: lib/src/graph/runtime_graph_reconstruction.dart
import '../reconstruction/reconstruct_graph.dart';
import 'runtime_graph.dart';

Map<String, dynamic> reconstructGraphFromIr(Map<String, dynamic> sources) {
  final graph = reconstructGraphFromSources(sources);
  return {
    'graph': graph.toJson(),
    'fingerprint': graphFingerprint(graph),
    'bounded': true,
  };
}

RuntimeGraph rebuildGraphFromPartial(RuntimeGraph partial) {
  final normalized = normalizeRuntimeGraph(partial);
  final nodes = <RuntimeNode>[];
  for (var i = 0; i < normalized.nodes.length; i++) {
    final n = normalized.nodes[i];
    nodes.add(RuntimeNode(
      id: n.id ?? 'rebuilt:$i',
      type: n.type,
      payload: n.payload,
    ));
  }
  return normalizeRuntimeGraph(
      RuntimeGraph(nodes: nodes, edges: normalized.edges));
}
